common/src/chainstate/utxo.rs

```diff
@@ -0,0 +1,462 @@
+//TODO: remove once the functions are used.
+#![allow(dead_code)]
+use crate::chain::{OutPoint, OutPointSourceId, Transaction, TxOutput};
+use crate::chainstate::Error;
+use crate::primitives::{BlockHeight, Id, Idable, H256};
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+
+use crate::chain::block::Block;
+use parity_scale_codec::{Decode, Encode};
+
+//todo: proper placement and derivation of this max
+const MAX_OUTPUTS_PER_BLOCK: u32 = 500;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct OutPointKey {
+    outpoint_hash: H256,
+    is_block_reward: bool,
+    index: u32,
+}
+
+impl OutPointKey {
+    pub fn outpoint_hash(&self) -> &H256 {
+        &self.outpoint_hash
+    }
+}
+
+impl From<&OutPoint> for OutPointKey {
+    fn from(outpoint: &OutPoint) -> Self {
+        let is_block_reward = match outpoint.get_tx_id() {
+            OutPointSourceId::Transaction(_) => false,
+            OutPointSourceId::BlockReward(_) => true,
+        };
+
+        let outpoint_hash =  match &outpoint.get_tx_id() {
+            OutPointSourceId::Transaction(inner) => inner.get(),
+            OutPointSourceId::BlockReward(inner) => inner.get(),
+        };
+
+        Self {
+            outpoint_hash,
+            is_block_reward,
+            index: outpoint.get_output_index(),
+        }
+    }
+}
+
+impl From<&OutPointKey> for OutPoint {
+    fn from(key: &OutPointKey) -> Self {
+        let id = if key.is_block_reward {
+            let utxo_id: Id<Block> = Id::new(&key.outpoint_hash);
+            OutPointSourceId::BlockReward(utxo_id)
+        } else {
+            let utxo_id: Id<Transaction> = Id::new(&key.outpoint_hash);
+            OutPointSourceId::Transaction(utxo_id)
+        };
+
+        OutPoint::new(id, key.index)
+    }
+}
+
+/// The Unspent Transaction Output
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct Utxo {
+    output: TxOutput,
+    is_block_reward: bool,
+    /// At which height this containing tx was included in the active block chain
+    height: BlockHeight,
+}
+
+impl Utxo {
+    pub fn new(output: TxOutput, is_block_reward: bool, height: BlockHeight) -> Self {
+        Self {
+            output,
+            is_block_reward,
+            height,
+        }
+    }
+
+    pub fn is_block_reward(&self) -> bool {
+        self.is_block_reward
+    }
+
+    pub fn height(&self) -> BlockHeight {
+        self.height
+    }
+
+    pub fn output(&self) -> &TxOutput {
+        &self.output
+    }
+
+    pub fn block_reward(&mut self, value: bool) {
```

Comment:
I don't think we need setters for this struct.

Reply:
Also, even if we do, this should be called "set_block_reward"

Reply:
removed.